include/eosio/bitset.hpp

```diff
@@ -0,0 +1,194 @@
+/**
+ *  @file
+ *  @copyright defined in eos/LICENSE
+ */
+#pragma once
+
+#include "from_bin.hpp"
+#include "from_json.hpp"
+#include "to_bin.hpp"
+#include "to_json.hpp"
+
+namespace eosio {
+
+// -------------------------------------------------------------------------------
+//      see https://github.com/AntelopeIO/spring/wiki/ABI-1.3:-bitset-type
+// -------------------------------------------------------------------------------
+
+
+// stores a bitset in a std::vector<uint8_t>
+//
+// - bits 0-7 in first byte, 8-15 in second, ...
+// - least significant bit of byte 0 is bit 0 of bitset.
+// - unused bits must be zero.
+// ---------------------------------------------------------------------------------
+struct bitset {
+   using buffer_type                         = std::vector<uint8_t>;
+   using size_type                           = uint32_t;
+   static constexpr size_type bits_per_block = 8;
+   static constexpr size_type npos           = static_cast<size_type>(-1);
+
+   static constexpr size_type calc_num_blocks(size_type num_bits) {
+      return (num_bits + bits_per_block - 1) / bits_per_block;
+   }
+
+   static size_type block_index(size_type pos) noexcept { return pos / bits_per_block; }
+   static uint8_t   bit_index(size_type pos)   noexcept { return static_cast<uint8_t>(pos % bits_per_block); }
+   static uint8_t   bit_mask(size_type pos)    noexcept { return uint8_t(1) << bit_index(pos); }
+
+   size_type size() const { return m_num_bits; }
+
+   void resize(size_type num_bits) {
+      m_bits.resize(calc_num_blocks(num_bits), 0);
+      m_num_bits = num_bits;
+      zero_unused_bits();
+   }
+
+   void clear() {
+      for (auto& byte : m_bits)
+         byte = 0;
+   }
+
+   void set(size_type pos) {
+      assert(pos < m_num_bits);
+      m_bits[block_index(pos)] |= bit_mask(pos);
+   }
+
+   bool operator[](size_type pos) const {
+      assert(pos < m_num_bits);
+      return !!(m_bits[block_index(pos)] & bit_mask(pos));
+   }
+
+   void zero_unused_bits() {
+      assert (m_bits.size() == calc_num_blocks(m_num_bits));
+
+      // if != 0 this is the number of bits used in the last block
+      const size_type extra_bits = bit_index(size());
+
+      if (extra_bits != 0)
+         m_bits.back() &= (uint8_t(1) << extra_bits) - 1;
+   }
+
+   bool unused_bits_zeroed() const {
+      // if != 0 this is the number of bits used in the last block
+      const size_type extra_bits = bit_index(size());
+      return extra_bits == 0 || (m_bits.back() & ~((uint8_t(1) << extra_bits) - 1)) == 0;
+   }
+
+   friend auto operator<=>(const bitset& a, const bitset& b) {
+      return std::tie(a.m_num_bits, a.m_bits) <=> std::tie(b.m_num_bits, b.m_bits);
+   }
+
+   friend bool operator==(const bitset& a, const bitset& b)  {
+      return std::tie(a.m_num_bits, a.m_bits) == std::tie(b.m_num_bits, b.m_bits);
+   }
```

Comment:
Any reason not to just use the default for `==` and `<=>` ?

Reply:
For the `to_key` implementation I need this specific order.
I could flip the order of the two members and use `= default`,  but because the order is important I thought better to have it specified. Probably could use a comment at least.

Reply:
https://github.com/AntelopeIO/abieos/pull/47/commits/b4fb1e40d500d2edf9d156e8c739d4ebc38e8901